akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala

```diff
@@ -0,0 +1,257 @@
+
+    builder.make()
+  }
+  val store: HTreeMap[String, ReplicatedData] =
+    db.hashMap("store")
+      .keySerializer(Serializer.STRING)
+      .valueSerializer(new MapdbReplicatedDataSerializer(context.system))
+      .createOrOpen()
+
+  val uncommitted = new java.util.HashMap[String, ReplicatedData]
+
+  override def postRestart(reason: Throwable): Unit = {
+    super.postRestart(reason)
+    // Load is only done on first start, not on restart
+    context.become(active)
+  }
+
+  override def postStop(): Unit = {
+    super.postStop()
+    commit()
+    db.close()
+  }
+
+  def receive = init
+
+  def init: Receive = {
+    case Load ⇒
+      try {
+        val loadData = LoadData(store.entrySet.asScala.map { entry ⇒
+          entry.getKey → entry.getValue
+        }(breakOut))
+        if (loadData.data.nonEmpty)
+          sender() ! loadData
+        sender() ! LoadCompleted
+        context.become(active)
+      } catch {
+        case NonFatal(e) ⇒
+          throw new LoadFailed("failed to load durable distributed-data", e)
+      }
+  }
+
+  def active: Receive = {
+    case Store(key, data, reply) ⇒
+      try {
+        if (commitInterval.length == 0) {
+          // optimize calls from put, no need to deserialize old value, using ThreadLocal
+          MapdbReplicatedDataSerializer.callFromPut.set(MapdbReplicatedDataSerializer.NotUsedOldValue)
+          try store.put(key, data) finally
+            MapdbReplicatedDataSerializer.callFromPut.set(null)
+          db.commit()
+        } else {
+          if (uncommitted.isEmpty)
+            context.system.scheduler.scheduleOnce(commitInterval, self, Commit)(context.system.dispatcher)
+          uncommitted.put(key, data)
+        }
+        reply match {
+          case Some(StoreReply(successMsg, _, replyTo)) ⇒
+            replyTo ! successMsg
+          case None ⇒
+        }
+      } catch {
+        case NonFatal(e) ⇒
+          log.error(e, "failed to store [{}]", key)
+          reply match {
+            case Some(StoreReply(_, failureMsg, replyTo)) ⇒
+              replyTo ! failureMsg
+            case None ⇒
+          }
+      }
+
+    case Commit ⇒
+      commit()
+  }
+
+  def commit(): Unit = {
+    if (!uncommitted.isEmpty()) {
+      val t0 = System.nanoTime()
+      try {
+        // optimize calls from put, no need to deserialize old value, using ThreadLocal
+        MapdbReplicatedDataSerializer.callFromPut.set(MapdbReplicatedDataSerializer.NotUsedOldValue)
+        try store.putAll(uncommitted) finally
+          MapdbReplicatedDataSerializer.callFromPut.set(null)
+        db.commit()
+        if (log.isDebugEnabled)
+          log.debug("store and commit of [{}] entries took [{} ms]", uncommitted.size,
+            TimeUnit.NANOSECONDS.toMillis(System.nanoTime - t0))
+      } catch {
+        case NonFatal(e) ⇒
+          import scala.collection.JavaConverters._
+          log.error(e, "failed to store [{}]", uncommitted.keySet.asScala.mkString(","))
```

Comment:
maybe say explicitly "store and/or commit". Maybe even better to log the two errors separately.